llama_hub/llama_packs/self_rag/base.py

```diff
@@ -0,0 +1,286 @@
+
+    Returns:
+        float: support score
+    """
+    isSup_score = 0
+    groundness_token_appear_id = -1
+    for tok_idx, token in enumerate(pred_tokens):
+        if token in _GROUND_TOKENS:
+            groundness_token_appear_id = tok_idx
+            break
+    if groundness_token_appear_id > -1:
+        grd_score_dict = {}
+        for token in _GROUND_TOKENS:
+            prob = pred_log_probs_dict[groundness_token_appear_id][token]
+            grd_score_dict[token] = np.exp(float(prob))
+        isSup_score = (
+            grd_score_dict["[Fully supported]"]
+            + 0.5 * grd_score_dict["[Partially supported]"]
+        ) / np.sum(list(grd_score_dict.values()))
+    return isSup_score
+
+
+def _is_useful_score(
+    pred_tokens: List[int], pred_log_probs_dict: List[Dict[str, float]]
+) -> float:
+    """Compute usefulness score
+
+    Args:
+        pred_tokens (List[int]): List of predicted tokens
+        pred_log_probs_dict (List[Dict[str, float]]): log probabilities of tokens for each predicted tokens
+
+    Returns:
+        float: relevance score
+    """
+    isUse_score = 0
+    utility_token_appear_id = -1
+    for tok_idx, tok in enumerate(pred_tokens):
+        if tok in _UTILITY_TOKENS:
+            utility_token_appear_id = tok_idx
+    if utility_token_appear_id > -1:
+        ut_score_dict = {}
+        for token in _UTILITY_TOKENS:
+            prob = pred_log_probs_dict[utility_token_appear_id][token]
+            ut_score_dict[token] = np.exp(float(prob))
+
+        ut_sum = np.sum(list(ut_score_dict.values()))
+        ut_weights = [-1, -0.5, 0, 0.5, 1]
+        isUse_score = np.sum(
+            [
+                ut_weights[i] * (ut_score_dict["[Utility:{}]".format(i + 1)] / ut_sum)
+                for i in range(len(ut_weights))
+            ]
+        )
+    return isUse_score
+
+
+class SelfRAGQueryEngine(CustomQueryEngine):
+    """Simple short form self RAG query engine."""
+
+    llm: Any = Field(default=None, description="llm")
+    retreiver: BaseRetriever = Field(default=None, description="Retreiver")
+    generate_kwargs: Dict = Field(default=None, description="llm generation arguments")
+    verbose: bool = Field(default=True, description="Verbose.")
+
+    def __init__(
+        self,
+        model_path: str,
+        retreiver: BaseRetriever,
+        verbose: bool = False,
+        model_kwargs: Dict = None,
+        generate_kwargs: Dict = None,
+        **kwargs: Any,
+    ) -> None:
+        """Init params."""
+        super().__init__(verbose=verbose, **kwargs)
+        model_kwargs = model_kwargs or _MODEL_KWARGS
+        self.generate_kwargs = generate_kwargs or _GENERATE_KWARGS
+        try:
+            from llama_cpp import Llama  # noqa: F401
+        except ImportError:
+            raise ImportError(_IMPORT_ERROR_MSG)
+        self.llm = Llama(model_path=model_path, verbose=verbose, **model_kwargs)
```

Comment:
we do have a llamacpp wrapper in llamaindex. out of curiosity is there a reason you are using the llama_cpp lib directly? (it's fine either way, just curious)

Reply:
oh nvm seems like its for logprobs